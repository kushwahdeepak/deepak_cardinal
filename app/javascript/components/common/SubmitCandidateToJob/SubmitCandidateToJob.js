import React, { useState } from "react";
import Form from "react-bootstrap/Form";
import Modal from "react-bootstrap/Modal";
import Button from "react-bootstrap/Button";
import Card from "react-bootstrap/Card";
import Alert from "react-bootstrap/Alert";
import Autocomplete from "react-autocomplete";
import axios from "axios";

function SubmitCandidateToJob({
  candidate,
  jobs,
  submissions,
  currentUser,
  show,
  onHide,
  setCandidateSubmited,
}) {
  const [selectJob, setSelectJob] = useState("");
  const [errorSubmitting, setErrorSubmitting] = useState(null);

  let jobItems = jobs && jobs.map((j) => {
    return {
      id: j[0],
      label: j[1],
      company: j[2],
      searchString: `${j[1]}(${j[2]}, jobid=${j[0]})`,
    };
  });

  const matchStateToTerm = (state, term) => {
    return state.searchString.toLowerCase().indexOf(term.toLowerCase()) !== -1;
  };

  const handleSubmit = () => {
    const job = jobItems.find((j) => j.searchString === selectJob);
    if (job == null || selectJob === "") return;

    var token = document.querySelector('meta[name="csrf-token"]').content;
    axios
      .post(`/submissions`, {
        authenticity_token: token,
        job_id: job.id,
        user_id: currentUser.id,
        person_id: candidate.id,
        commit: "Submit",
      })
      .then((res) => {
        onHide();
        setCandidateSubmited(true);
        setSelectJob("");
        setErrorSubmitting(null);
      })
      .catch((err) => {
        setErrorSubmitting(err.message);
      });
  };

  const hideModal = () => {
    setSelectJob("");
    setErrorSubmitting(null);
    onHide();
  };

  return (
    <Modal
      onHide={hideModal}
      show={show}
      size="lg"
      aria-labelledby="contained-modal-title-vcenter"
      scrollable
    >
      <Modal.Header closeButton>
        <Modal.Title id="contained-modal-title-vcenter" className="mx-auto">
          Submit candidate to a job
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <Card className="my-3">
          <Card.Body>
            {errorSubmitting && (
              <Alert
                variant="danger"
                onClose={() => setErrorSubmitting(null)}
                dismissible
              >
                {errorSubmitting}
              </Alert>
            )}
            <Form>
              <Form.Group>
                <Form.Label>Job</Form.Label>
                <Autocomplete
                  className="w-100"
                  shouldItemRender={matchStateToTerm}
                  getItemValue={(item) => item.searchString}
                  items={jobItems}
                  renderItem={(item, isHighlighted) => (
                    <div
                      key={item.id}
                      style={{
                        background: isHighlighted ? "lightgray" : "white",
                      }}
                    >
                      {item.searchString}
                    </div>
                  )}
                  renderInput={(props) => {
                    return <input {...props} style={{ width: "100%" }} />;
                  }}
                  wrapperStyle={{ display: "block" }}
                  value={selectJob}
                  onChange={(e, v) => setSelectJob(v)}
                  onSelect={(val) => setSelectJob(val)}
                />
              </Form.Group>
            </Form>
            <hr />
            <Card.Title>Existing Submissions</Card.Title>

            {submissions && submissions.map((s, index) => (
              <Card.Body key={index}> {s} </Card.Body>
            ))}
          </Card.Body>
        </Card>
      </Modal.Body>
      <Modal.Footer>
        <Button variant="light" onClick={hideModal}>
          Close
        </Button>
        <Button variant="primary" onClick={handleSubmit}>
          Submit
        </Button>
      </Modal.Footer>
    </Modal>
  );
}

export default SubmitCandidateToJob;
