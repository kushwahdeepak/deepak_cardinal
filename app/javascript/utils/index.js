import { act } from 'react-dom/test-utils'

export const findByTestAttr = (wrapper, attr) => {
    return wrapper.find(`[data-test='${attr}']`)
}

export const timeoutResolve = (time = 0, response = {}) =>
    new Promise((resolve) => {
        setTimeout(() => resolve(response), time)
    })

export const wait = (time = 0) => {
    const timer = timeoutResolve(time)

    return act(async () => {
        await timer
    })
}
export function emailIsValid(email) {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)
}

export function generateRandom(min, max) {
    return Math.floor(min + Math.random() * (max + 1 - min))
}

export function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase()
}

export function createMarkedList(array) {
    if (!Array.isArray(array)) {
        return []
    }
    return array.map((candidate) => ({
        id: candidate.id,
        name: candidate.id,
        fullName: `${candidate.first_name} ${candidate.last_name}`,
        emailAddress: candidate.email_address,
        isChecked: false,
    }))
}

export function stringReplace(string,toSymbol,fromSymbol){
    return string.replace(toSymbol, fromSymbol)
}

export function capitalize(string) {
    return string.replace(/\b(\w)/g, s => s.toUpperCase());
}

export function skillIsNotEmpty(value){
    return value.trim().length > 0;
}

export function firstCharacterCapital(value){
    return value?.charAt(0)?.toUpperCase() + value?.slice(1)
}
