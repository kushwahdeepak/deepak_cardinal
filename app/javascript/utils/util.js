import isNil from 'lodash.isnil'
import isEmpty from 'lodash.isempty'

class Util {
    /**
     * Define stage values
     */
    static STAGES = ['first_interview', 'second_interview', 'offer', 'reject']

    /**
     * Tests whether a single object attribute includes the at
     * least one string from the given set.
     */
    static attributePassesStringSetFilter(object, attr, stringSet) {
        if (isNil(object[attr])) return false

        // We "OR" the tag tests: the attribute needs to contain only one tag
        return stringSet
            .map((string) => string.trim().toLowerCase())
            .reduce(
                (passes, string) =>
                    passes ||
                    Util.attributeToString(object[attr])
                        .toLowerCase()
                        .includes(string),
                false
            )
    }

    /**
     * Tests whether at least one of a set of object attributes
     * includes every word from a given piece of text.
     */
    static objectPassesTextFilter(object, attrs, text) {
        text = text.trim().toLowerCase()

        if (text === '') return true

        const words = text.split(' ')

        // Join all the given attribute names to lowercase, stringified attribute values
        const allAttrsString = attrs
            .filter((attr) => !isNil(object[attr]))
            .map((attr) => Util.attributeToString(object[attr]).toLowerCase())
            .join('')

        // Is every word included in the allAttrsString?
        return words.reduce(
            (matchesAll, word) => matchesAll && allAttrsString.includes(word),
            true
        )
    }

    static attributeToString(attr) {
        if (typeof attr === 'string') {
            return attr
        } else if (Array.isArray(attr)) {
            return attr.join(',')
        } else if (typeof attr === 'object') {
            return JSON.stringify(attr)
        } else {
            return '' + attr
        }
    }

    static numberWithCommas(x) {
        return x?.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
    }

    /**
     * Returns number of results in format [x] - [y] of [z]
     * @param {number} totalJobsCount Total number of jobs
     * @param {number} pageCount Total number of pages
     * @param {number} activePage Currently active page
     * @param {number} pageSize How many results to display per page
     */
    static displayNumberOfResults(
        totalJobsCount,
        pageCount,
        activePage,
        pageSize,
        totalPersonCount=null,
        candidatePage=false,
        showdisplaycount=true
    ) {
        if (totalJobsCount == 0) return '0 - 0 of 0'
        const offset = pageSize * pageCount - totalJobsCount
        const lastJobIndex =
            (activePage + 1) * pageSize > totalJobsCount
                ? totalJobsCount
                : (activePage + 1) * pageSize
        const firstJobIndex =
            activePage == 0
                ? 1
                : activePage == pageCount - 1
                ? lastJobIndex - pageSize + offset + 1
                : lastJobIndex - pageSize + 1
        return `${firstJobIndex} - ${lastJobIndex} of ${Util.numberWithCommas(
            (candidatePage && showdisplaycount) ? totalPersonCount : totalJobsCount

        )}`
    }

    /**
     * Returns a string indicating the candidate's title and current
     * company, based on available information on the Person object
     * @param {object} candidate Person object from our db
     */
    static candidateCompanyString(candidate) {
        const position = candidate.company_position
        const currentCompany = candidate.current_company

        if (isEmpty(position) && isEmpty(currentCompany)) return ''

        if (!isEmpty(position) && !isEmpty(currentCompany))
            return `${position} at ${currentCompany}`

        if (!isEmpty(position)) return position

        if (!isEmpty(currentCompany)) return `at ${currentCompany}`
    }

    /**
     * Converts the given array of objects with ids to an object whose
     * keys are the ids of the array objects.
     *
     * Every object in the given array must have a unique 'id' property defined.
     *
     * @param {Array} array
     * @returns {Object}
     */
    static arrayToObject(array) {
        const obj = {}
        array.forEach((element) => (obj[element.id] = element))
        return obj
    }

    /**
     * Makes a copy of the given array and replace the value at 'index' with 'value'
     *
     * @param {Array} array
     * @param {Number} index
     * @param {*} value
     * @returns {Array}
     */
    static replaceInArrayCopy_i(array, index, value) {
        const clone = array.slice()
        clone[index] = value
        return clone
    }

    /**
     * Makes a copy of the given array and replaces the element with the same
     * .id property as obj.id with obj
     *
     * @param {Array} array
     * @param {Object} obj
     * @returns {Array}
     */
    static replaceInArrayCopy(array, obj) {
        const index = array.findIndex((element) => element.id === obj.id)
        return Util.replaceInArrayCopy_i(array, index, obj)
    }

    static deleteInArrayCopy(array, index) {
        const copy = array.slice()
        copy.splice(index, 1)

        return copy
    }

    static handleUndefined(value){
        if (value === undefined || value === 'undefined')
        return '';
        else
        return value
    }

    static handleUndefinedFullName(firstName, lastName){
      if(firstName === "undefined" || firstName === undefined)
      return lastName
      else if(lastName === "undefined" || lastName === undefined)
      return firstName
      else
      return firstName + ' ' + lastName     }

    static capitalize(value){
         return value.charAt(0).toUpperCase() + value.slice(1).toLowerCase();
    }
}

export default Util
