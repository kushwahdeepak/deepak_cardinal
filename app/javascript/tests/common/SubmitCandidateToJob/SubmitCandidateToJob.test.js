import React from 'react';
import SubmitCandidateToJob from '../../../components/common/SubmitCandidateToJob/SubmitCandidateToJob';
import renderer from 'react-test-renderer';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { findByTestAttr } from '../../../utils';

configure({adapter: new Adapter()});

const initialShallowRender = (props = {}) => {
	return shallow(<SubmitCandidateToJob {...props}/>);
};

const initialMountRender = (props = {}) => {
	return mount(<SubmitCandidateToJob {...props}/>);
};


describe('SubmitCandidateToJob component',  () => {
	let wrapper;
	let fullWrapper;

	beforeEach(() => {

		const props = {
			jobs: [
				{
					id: 'MockId',
					label: 'Mock label',
					company: 'Mock Company',
					searchString: 'Mock Search String'
				}
			]
		};
		wrapper = initialShallowRender(props);
		fullWrapper = initialMountRender(props);
	});

	afterEach(() => {
		fullWrapper.unmount();
	});

	it('should render SubmitCandidateToJob component', () => {
		expect(wrapper.exists()).toBe(true);
	});

	it('should match with a snapshot', () => {
		const component = renderer.create(<SubmitCandidateToJob />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

});
