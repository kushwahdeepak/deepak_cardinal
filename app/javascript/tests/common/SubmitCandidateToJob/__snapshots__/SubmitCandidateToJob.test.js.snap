// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitCandidateToJob component should match with a snapshot 1`] = `null`;
